### PR TITLE
WIP: specialized Ace highlighting/mode for custom expression

### PR DESCRIPTION
**Pull request by ariya**
1. Ask a question, Custom questioh
2. Sample Dataset, People table
3. Filter, Custom Expression and type `NOT lower([State]) = "ca"`

**Before**

image

**After**

image

### CODE REVIEW

**Comment by gusaiani** [comment]
@ariya Did you want a deep review at this point? :))

Syntax highlight is definitely working, pretty awesome.

It isn't exactly what we currently have in terms of colors, it looks more granular than the original (NOT is highlighted now, isn't originally).

Happy to see colors here, even with the way they are syntaxed. And happy to play around more, whatever helps.

**Comment by ariya** [comment]
> @ariya Did you want a deep review at this point? :))

This is going to the feature branch anyway, so maybe just  a sanity check for now.

> Syntax highlight is definitely working, pretty awesome.
> It isn't exactly what we currently have in terms of colors, it looks more granular than the original (NOT is highlighted now, isn't originally).

Yeah, the colors are off but hopefully you will be able help out with custom CSS/theme as a possible follow-up tweaks.

File: frontend/src/metabase/query_builder/components/expressions/ExpressionMode.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+          token: "string",
+          regex: '".*?"',
+        },
+        {
+          token: "constant.numeric",
+          regex: "[+-]?\\d+(?:(?:\\.\\d*)?(?:[eE][+-]?\\d+)?)?\\b",
+        },
+        {
+          token: "variable",
+          regex: "[[a-zA-Z0-9_s→]+\\]",
```
Comment (flamber) [inline frontend/src/metabase/query_builder/components/expressions/ExpressionMode.js:21]: This makes me feel like it doesn't like other languages.
But it also doesn't like spaces, or dashes.
image

File: frontend/src/metabase/query_builder/components/expressions/ExpressionMode.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  "ace/mode/text_highlight_rules",
+).TextHighlightRules {
+  constructor() {
+    super();
+
+    this.$rules = {
+      start: [
+        {
+          token: "string",
+          regex: '".*?"',
```
Comment (flamber) [inline frontend/src/metabase/query_builder/components/expressions/ExpressionMode.js:13]: I know single-quotes are converted onblur, but still
image

**Comment by gusaiani** [comment]
> Yeah, the colors are off but hopefully you will be able help out with custom CSS/theme as a possible follow-up tweaks.

Great, happy to colorize all the things.

File: frontend/src/metabase/query_builder/components/expressions/ExpressionMode.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+class ExpressionHighlight extends window.ace.acequire(
+  "ace/mode/text_highlight_rules",
+).TextHighlightRules {
+  constructor() {
+    super();
+
+    this.$rules = {
+      start: [
+        {
+          token: "string",
```
Comment (flamber) [inline frontend/src/metabase/query_builder/components/expressions/ExpressionMode.js:12]: I think it's going to slightly bork, when people are using regex - probably not a huge problem - example:
image

**Review by flamber — APPROVED** [review]
Better than currently (no colors), so :+1:

File: frontend/src/metabase/query_builder/components/expressions/ExpressionMode.js (hunk 1 above)
Comment (ariya) [reply frontend/src/metabase/query_builder/components/expressions/ExpressionMode.js]: I've tweaked this one, so anything between the square brackets should be recognized properly.

File: frontend/src/metabase/query_builder/components/expressions/ExpressionMode.js (hunk 2 above)
Comment (ariya) [reply frontend/src/metabase/query_builder/components/expressions/ExpressionMode.js]: Should be fixed now, thanks!

File: frontend/src/metabase/query_builder/components/expressions/ExpressionMode.js (hunk 3 above)
Comment (ariya) [reply frontend/src/metabase/query_builder/components/expressions/ExpressionMode.js:16]: Likewise!

**Pull request merged**